### DISCUSSION TITLE
Pactum and Fastify

### DISCUSSION THREAD

**Discussion by leegee** [pactumjs/pactum#101] [Q&A] [answered]
Is it possible to use Pactum against a Fastify server, in the manner of Supertest or the Fastify manual-recommended way, without having the server running? Currently ever test file is starting a server, and when run in parallel, of course the port is taken. Varying the server port seems the wrong with to go.

**Comment by leelaprasadv** [comment, marked as answer]
Hi 👋,

Can you let me know which test runner are you using? 
Does your setup require you to run server instances for every test? If not (assuming you might be using mocha), you can start your fastify server in a `before()` and stop it in an `after()` global hooks of mocha. This way you just have to run one instance of your server before all tests begin on a single port.

Ex: add a `base.spec.js`
```js
const server = require('./path-to-fastify-server');

// global hooks
before( () => {
  server.start(); // start fastify server before tests
});

after( () => {
  server.stop(); // if implemented for graceful shutdown
});
```

Hope this helps! 

### Edited (07/02/22)

Use this package - https://www.npmjs.com/package/pactum-supertest

**Reply by leegee** [reply]
Hi, and thanks for the answer. 

What I'm looking for is to execute Pactum tests against a non-running server, as one does with superagent/supertest. Doing so saves a lot of run time, and increases isolation.

Is this possible or will it be in the future?

Pactum has a lot going for it in terms of brevity and clarity, but this is the sticking point that has prevented me recommending it to clients.

**Reply by ASaiAnudeep** [reply]
> What I'm looking for is to execute Pactum tests against a non-running server, as one does with superagent/supertest. Doing so saves a lot of run time, and increases isolation.

supertest runs the tests against a running server. Usually it [starts](https://github.com/visionmedia/supertest/blob/master/lib/test.js#L61) and [stops](https://github.com/visionmedia/supertest/blob/master/lib/test.js#L117-L118) the server for each and every test. I'm not sure how this will enhance the testing experience.

> Varying the server port seems the wrong with to go.

This is essentially what supertest does in the background - *running the server in random port for every test*

> Is this possible or will it be in the future?

We are discussing this internally to understand the efforts we need to put to achieve similar functionality. I never got a chance to work on supertest. I would love to hear more from you about the advantages of having such behaviour. One advantage I can clearly see is running parallel tests (*by running the server on random ports*).

**Reply by leegee** [reply]
Yes, it's the parallel execution that is valuable, not just for speed but for simplicity of isolation.  

The server start-up/shut-down process is also a lot cleaner with supertest --- yes, I know the example you gave is really neat and clean, but the clients I deal with are often very hard to shift to new tech, for no good reason... 

Also, thank you very much for Pactum: I have found it a joy to work with. If you could publish something on writing plugins...that'd be very helpful in the future!

**Reply by ASaiAnudeep** [reply]
What test runner do you use for running tests?

**Reply by leegee** [reply]
> What test runner do you use for running tests?

Mocha or Jest.

**Reply by ASaiAnudeep** [reply]
@leegee Thanks for being patient. I have written and published a small plugin that will behave similar to `supertest`.  Please give it a try and let us know the feedback.

Package - https://www.npmjs.com/package/pactum-supertest

**Reply by leegee** [reply]
Very nice, thank you -- having something published reassures some clients, and saves a lot of cut'n'paste and all the badness that entails.

A couple of points, if I may:

Might it be an idea to allow testing of hosts other than `localhost`?

Might it be safer to test for the test-runner hooks? 

```
  if (typeof before === 'function') {
    before(start);
    after(stop);
  }

  else if (typeof beforeAll === 'function') {
    beforeAll(start);
    afterAll(stop);
  }
```

Or maybe just export `start/stop` so users can call them as needed (`beforeEach`, etc)? IDK

Again, thank you for your work.

**Reply by ASaiAnudeep** [reply]
Ahh.. this makes more sense. Thanks for your suggestions. Created issues to track them.

https://github.com/pactumjs/pactum-supertest/issues/1
https://github.com/pactumjs/pactum-supertest/issues/2
https://github.com/pactumjs/pactum-supertest/issues/3
